### PR TITLE
Add Julia port of cbrt from Openlibm.

### PR DESCRIPTION
**Pull request by pkofod**
<s>*A little note: some functions/methods such as, `issubnormal` and more should probably be in `Base.Math`, but for now I'm placing the methods that I use with the methods. When all the openlibm stuff is out of `Base.Math` I will reorganize this slightly. Focus is on getting the rest of the functions in there asap focusing only on correctness and performance. If this is not wanted, we can look at code organization (inside of the module mind you), but my time is limited these days.*</s>

I would love advice from @stevengj and @simonbyrne on good test values. I've basically grabbed some a bit out of thin air.

I'm doing some preliminary benchmarks, but of course nanosoldier  should receive benchmark suites for these functions.

*EDIT*
Some relative timings. The sampled values are
```
[   4.94066e-324
       9.88131e-324
       1.4822e-323 
      -1.4822e-323 
      -9.88131e-324
      -4.94066e-324
       0.45        
       0.6         
       0.98        
       1.0         
  157464.0         
       1.22504e6   
       4.096e6     
       9.6636e6    
       1.88211e7   
       3.24618e7   
       5.14788e7   
       7.67656e7   
       1.09215e8   
       1.49721e8   
       1.99177e8   
       2.58475e8   
       3.28509e8   
       4.10172e8   
       5.04358e8   
       6.1196e8    
       7.33871e8   
       8.70984e8   
       1.02419e9   
      -1.0         
 -157464.0         
      -1.22504e6   
      -4.096e6     
      -9.6636e6    
      -1.88211e7   
      -3.24618e7   
      -5.14788e7   
      -7.67656e7   
      -1.09215e8   
      -1.49721e8   
      -1.99177e8   
      -2.58475e8   
      -3.28509e8   
      -4.10172e8   
      -5.04358e8   
      -6.1196e8    
      -7.33871e8   
      -8.70984e8   
      -1.02419e9   
      -1.79769e308 
       1.79769e308]
```
But I've left out the x-values because it only made the plot worse to look at. It's in the same order as what you see above though.

**Float32**
fl32

**Float64**
fl64

So, I'm basically seeing either noise, or a 0.93 ratio (master/v0.6)

### CODE REVIEW

**Comment by ViralBShah** [comment]
Love all these! Would be great to be able to remove openlibm completely from our distribution.

File: base/special/cbrt.jl [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        u = reinterpret(UInt64, t)
+        u = (u + 0x80000000) & UInt64(0xffffffffc0000000)
+        t = reinterpret(Float64, u)
+
+        # one step Newton iteration to 53 bits with error < 0.667 ulps
+        s = t*t             # t*t is exact */
+        r = x/s             # error <= 0.5 ulps; |r| < |t| */
+        w = t + t           # t+t is exact */
+        r = (r - t)/(w + r) # r-t is exact; w+r ~= 3*t */
+        t = t + t*r         # error <= 0.5 + 0.5/3 + epsilon */
```
Comment (stevengj) [inline base/special/cbrt.jl]: `muladd`?

File: base/special/cbrt.jl [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    # almost exactly the RHS of the above provided we first subtract the
+    # exponent bias and later add it back.  We do the
+    # subtraction virtually to keep e >= 0 so that ordinary integer
+    # division rounds towards minus infinity; this is also efficient.
+
+    if issubnormal(x, hw) # zero or subnormal?
+        if x == zero(Tf)
+            return x # cbrt(+-0) is itself
+        end
+        t = x*cbrt_t(Tf)
```
Comment (stevengj) [inline base/special/cbrt.jl]: Couldn't this overflow if `x` is close to `realmax(Tf)`?

File: test/math.jl [hunk 3]
```diff
@@ -900,3 +900,29 @@ float(x::FloatWrapper) = x
...
+        @test cbrt(-T(Inf)) === -T(Inf)
+        @test isnan_type(T, cbrt(T(NaN)))
+        for x in (pcnfloat(nextfloat(nextfloat(zero(T))))...,
+                  pcnfloat(prevfloat(prevfloat(zero(T))))...,
+                  0.45, 0.6, 0.98,
+                  map(x->x^3, 1.0:1.0:1024.0)...,
+                  map(x->-x^3, 1.0:1.0:1024.0)...,
+                  nextfloat(-T(Inf)), prevfloat(T(Inf)))
+            by = cbrt(big(T(x)))
+            @test T((cbrt(T(x)) - by))/eps(abs(T(by))) <= 1
```
Comment (stevengj) [inline test/math.jl]: `@test cbrt(T(x)) ≈ by rtol=eps(T)` ?

File: test/math.jl [hunk 4]
```diff
@@ -900,3 +900,29 @@ float(x::FloatWrapper) = x
...
+        @test cbrt(-one(T)) === -one(T)
+        @test cbrt(T(Inf)) === T(Inf)
+        @test cbrt(-T(Inf)) === -T(Inf)
+        @test isnan_type(T, cbrt(T(NaN)))
+        for x in (pcnfloat(nextfloat(nextfloat(zero(T))))...,
+                  pcnfloat(prevfloat(prevfloat(zero(T))))...,
+                  0.45, 0.6, 0.98,
+                  map(x->x^3, 1.0:1.0:1024.0)...,
+                  map(x->-x^3, 1.0:1.0:1024.0)...,
+                  nextfloat(-T(Inf)), prevfloat(T(Inf)))
```
Comment (stevengj) [inline test/math.jl]: Since you do the `bym` test, all these negative test values seem redundant?

File: base/special/cbrt.jl (hunk 2 above)
Comment (pkofod) [reply base/special/cbrt.jl]: sure, but `x` is subnormal here

File: test/math.jl (hunk 4 above)
Comment (pkofod) [reply test/math.jl]: ah, that's right!

File: test/math.jl (hunk 3 above)
Comment (pkofod) [reply test/math.jl]: ah, alright, I didn't know these syntaxes (syntaces?) for `at-test`

File: base/special/cbrt.jl [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+cbrt_t(::Type{Float32}) = 1.6777216f7 ## 2f0^24
+
+cbrt_t_from_words(::Type{Float32}, s, high)   = reinterpret(Float32, s | (div(high & 0x7fffffff, 0x00000003))+cbrt_B2_32)
+cbrt_t_from_words_alt(::Type{Float32}, s, hx) = reinterpret(Float32, s | (div(hx, 0x00000003) + cbrt_B1_32))
+
+cbrt_t_from_words(::Type{Float64}, s, high)   = reinterpret(Float64, UInt64(s | (div(high & 0x7fffffff, 0x00000003)+cbrt_B2_64))<<32)
+cbrt_t_from_words_alt(::Type{Float64}, s, hx) = reinterpret(Float64, UInt64(s | (div(hx, 0x00000003) + cbrt_B1_64))<<32)
+
+function cbrt(x::Tf) where Tf <: Union{Float32, Float64}
+    # mathematically cbrt(x) is defined as the real number y such that y^2 == x
```
Comment (dpsanders) [inline base/special/cbrt.jl]: y^3

File: test/math.jl [hunk 6]
```diff
@@ -917,12 +917,13 @@ pcnfloat(x) = prevfloat(x), x, nextfloat(x)
                   pcnfloat(prevfloat(prevfloat(zero(T))))...,
                   0.45, 0.6, 0.98,
                   map(x->x^3, 1.0:1.0:1024.0)...,
-                  map(x->-x^3, 1.0:1.0:1024.0)...,
                   nextfloat(-T(Inf)), prevfloat(T(Inf)))
             by = cbrt(big(T(x)))
-            @test T((cbrt(T(x)) - by))/eps(abs(T(by))) <= 1
+            #@test T((cbrt(T(x)) - by))/eps(abs(T(by))) <= 1
```
Comment (stevengj) [inline test/math.jl]: Why is this comment here?

File: base/special/cbrt.jl (hunk 5 above)
Comment (pkofod) [reply base/special/cbrt.jl]: hah, well, that's embarrassing

File: test/math.jl (hunk 6 above)
Comment (pkofod) [reply test/math.jl]: I'll remove it. Are the tests equivalent, though? I Is the `rtol` thing testing that it is with 1 ulp of the correctly rounded result?

File: base/special/cbrt.jl [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+# Optimized by Bruce D. Evans.
+
+# s_cbrtf.c -- float version of s_cbrt.c.
+# Conversion to float by Ian Lance Taylor, Cygnus Support, ian@cygnus.com.
+# Debugged and optimized by Bruce D. Evans.
+
+cbrt(x::Real) = cbrt(float(x))
+
+issubnormal(::Float64, hw) = hw < 0x00100000
+issubnormal(::Float32, hw) = hw < 0x00800000
```
Comment (JeffBezanson) [inline base/special/cbrt.jl]: Base already exports an `issubnormal` that this should use.

File: base/special/cbrt.jl [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        u = reinterpret(UInt64, t)
+        u = (u + 0x80000000) & UInt64(0xffffffffc0000000)
+        t = reinterpret(Float64, u)
+
+        # one step Newton iteration to 53 bits with error < 0.667 ulps
+        s = t*t             # t*t is exact
+        r = x/s             # error <= 0.5 ulps; |r| < |t|
+        w = t + t           # t+t is exact
+        r = (r - t)/(w + r) # r-t is exact; w+r ~= 3*t
+        t = muladd(t*r, t)         # error <= 0.5 + 0.5/3 + epsilon
```
Comment (JeffBezanson) [inline base/special/cbrt.jl]: `muladd(t, r, t)` ? :)

File: base/special/cbrt.jl (hunk 7 above)
Comment (pkofod) [reply base/special/cbrt.jl]: I've used it. It does cost 0.3 ns more because I lumped 0 and subnormals before :)

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25948#issuecomment-365043507) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/f31728d3cbd837c4479ba2f6910562f153c0b655/5749345_vs_131813f/report.md). cc @ararslan

**Comment by pkofod** [comment]
@ararslan could you refire nanosoldier here? (or should I fix conflicts first?)

**Comment by ararslan** [comment]
Always best to fix conflicts before running Nanosoldier; when there are conflicts, Nanosoldier checks out the branch tip rather than the merge commit with master.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by ararslan** [comment]
Looks like I got ignored.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25948#issuecomment-365745701) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/dbc477d494745e2502b34327af3cd9a3ea883b04/e75cca5_vs_a2e4f16/report.md). cc @ararslan

**Comment by pkofod** [comment]
Anything I need to do to finish this up?

**Comment by stevengj** [comment]
Are the benchmark regressions unrelated?

**Comment by ararslan** [comment]
The ones reported by Nanosoldier appear to be.

**Comment by pkofod** [comment]
Hm, CI situation seems messy but unrelated.

**Comment by JeffreySarnoff** [comment]
> @inline highword(x::UInt64)  = (x >> 32) % UInt32

The shifting intent is better expressed `>>>`,  and precludes one future mistake.
This would not change current behavior.

**Comment by pkofod** [comment]
> The shifting intent is better expressed >>>, and precludes one future mistake.
> This would not change current behavior.

gotcha!

**Comment by pkofod** [comment]
I think this might be good to go, @simonbyrne @stevengj do you have any further comments?

**Comment by pkofod** [comment]
Nanosoldier can be activated again here.

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25948#issuecomment-389498123) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6a2692ad03804c2f28c7f76148a6b181c6af5097/cff470d_vs_6291d3e/report.md). cc @ararslan

File: LICENSE.md [hunk 9]
```diff
@@ -72,8 +72,12 @@ The following components of Julia's standard library have separate licenses:
 - base/sparse/cholmod.jl (see [SUITESPARSE](http://faculty.cse.tamu.edu/davis/suitesparse.html))
 - base/special/exp.jl (see [FDLIBM](http://www.netlib.org/fdlibm/e_exp.c) [Freely distributable with preserved copyright notice])
 - base/special/rem_pio2.jl (see [FDLIBM](http://www.netlib.org/fdlibm/e_rem_pio2.c) [Freely distributable with preserved copyright notice])
+<<<<<<< HEAD
```
Comment (KristofferC) [inline LICENSE.md]: ⚠️

**Review by KristofferC — DISMISSED** [review]

File: LICENSE.md (hunk 9 above)
Comment (pkofod) [reply LICENSE.md]: Bah, Atom's git(hub) plugin did not agree with the way I clicked it's buttons :) Also had some extra <<<<'s in one of the other PRs

**Comment by simonbyrne** [comment]
I've made a few changes which hopefully should make it a bit clearer (e.g. avoiding lots of magic constants, in favour of letting the constants be propagated at compile time). I don't see any performance regressions.

**Comment by pkofod** [comment]
> I've made a few changes which hopefully should make it a bit clearer (e.g. avoiding lots of magic constants, in favour of letting the constants be propagated at compile time). I don't see any performance regressions.

Good stuff!

**Comment by ararslan** [comment]
The 64-bit Linux failure on Travis is unrelated:
```
Worker 4 failed running test channels:
Some tests did not pass: 180 passed, 1 failed, 0 errored, 0 broken.channels: Test Failed at /tmp/julia/share/julia/test/channels.jl:261
  Expression: isempty(Base.Workqueue)
Stacktrace:
 [1] record(::Test.DefaultTestSet, ::Test.Fail) at /home/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/Test/src/Test.jl:741
 [2] (::getfield(Main, Symbol("##44#50")))() at /tmp/julia/share/julia/test/runtests.jl:256
 [3] cd(::getfield(Main, Symbol("##44#50")), ::String) at ./file.jl:72
 [4] top-level scope
 [5] include at ./boot.jl:314 [inlined]
 [6] include_relative(::Module, ::String) at ./loading.jl:1069
 [7] include(::Module, ::String) at ./sysimg.jl:29
 [8] exec_options(::Base.JLOptions) at ./client.jl:264
 [9] _start() at ./client.jl:424
```
and the 32-bit Windows build timed out. I'd say we're all good here. :slightly_smiling_face:

File: base/special/cbrt.jl [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+With the IEEE floating point representation, for finite positive normal values, ordinary
+integer divison of the value in bits magically gives almost exactly the RHS of the above
+provided we first subtract the exponent bias and later add it back.  We do the
+subtraction virtually to keep e >= 0 so that ordinary integer division rounds towards
+minus infinity; this is also efficient. All operations can be done in 32-bit.
+
+These implementations assume that NaNs, infinities and zeros have already been filtered.
+"""
+@inline function _approx_cbrt(x::T) where {T<:Union{Float32,Float64}}
+    adj = exponent_bias(T)*2/3 - 0.03306235651
```
Comment (stevengj) [inline base/special/cbrt.jl]: should this be computed in `T` precision?

File: base/special/cbrt.jl [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    k = significand_bits(T) - (8*sizeof(T) - 32)
+
+    u = highword(x) & 0x7fff_ffff
+    if u >= Base.Math.highword(realmin(T))
+        v = div(u, UInt32(3)) + floor(UInt32, adj * 2.0^k)
+    else
+        # subnormal
+        x *= maxintfloat(T)
+        adj -= exponent(maxintfloat(T))/3
```
Comment (stevengj) [inline base/special/cbrt.jl]: should we multiply by `T(0.333333333333333333)` here instead of dividing by `3`?

File: base/special/cbrt.jl [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+These implementations assume that NaNs, infinities and zeros have already been filtered.
+"""
+@inline function _approx_cbrt(x::T) where {T<:Union{Float32,Float64}}
+    adj = exponent_bias(T)*2/3 - 0.03306235651
+    k = significand_bits(T) - (8*sizeof(T) - 32)
+
+    u = highword(x) & 0x7fff_ffff
+    if u >= Base.Math.highword(realmin(T))
+        v = div(u, UInt32(3)) + floor(UInt32, adj * 2.0^k)
```
Comment (stevengj) [inline base/special/cbrt.jl]: `exp2`?

File: base/special/cbrt.jl [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    s = t*t             # t*t is exact
+    r = x/s             # error <= 0.5 ulps; |r| < |t|
+    w = t+t             # t+t is exact
+    r = (r - t)/(w + r) # r-t is exact; w+r ~= 3*t
+    t = muladd(t, r, t) # error <= 0.5 + 0.5/3 + epsilon
+    return t
+end
+
+function cbrt(x::Union{Float32,Float64})
+    if !isfinite(x) || x == zero(x)
```
Comment (stevengj) [inline base/special/cbrt.jl]: `iszero(x)`, though it probably makes no difference

Comment (simonbyrne) [reply base/special/cbrt.jl]: i forgot we had `iszero`

File: base/special/cbrt.jl (hunk 10 above)
Comment (simonbyrne) [reply base/special/cbrt.jl]: No, it is later converted to `UInt32`. These are also constants, so should be evaluated at compile time.

File: base/special/cbrt.jl (hunk 11 above)
Comment (simonbyrne) [reply base/special/cbrt.jl]: Again, is a constant, so should be evaluated at compile time.

File: base/special/cbrt.jl (hunk 12 above)
Comment (simonbyrne) [reply base/special/cbrt.jl]: fixed.

File: base/special/cbrt.jl (hunk 13 above)
Comment (StefanKarpinski) [reply base/special/cbrt.jl]: Writing `x + x === x` might be a tricksy way of writing this condition. Analysis:
* `-0.0 + -0.0 === -0.0`
* `0.0 + 0.0 === 0.0`
* `Inf + Inf === Inf`
* `-Inf + -Inf === -Inf`
* `NaN + NaN === NaN`

It should be false for any non-zero finite value. The last one is subtle one: the result of `+` when both arguments are NaN will be one of the arguments which are both bitwise identical to each other and thus to the result. I forget if this is guaranteed by the IEEE standard or not; I also forget if `===` requires moving to an integer register or if there's an fp operation for exact bit comparison. If not, then we could just do `isequal(x + x, x)` instead.

**Review by stevengj — APPROVED** [review]

**Comment by simonbyrne** [comment]
Okay, I've squashed the contributions into 2 commits. Ready to go.

**Comment by ViralBShah** [comment]
Probably should be rebased on master.

**Comment by simonbyrne** [comment]
I just did

**Comment by pkofod** [comment]
Great stuff! And have you verified that timing don't regress (too much) @simonbyrne ?

**Comment by ViralBShah** [comment]
What's causing the `SparseArrays` error in CI?

**Comment by fredrikekre** [comment]
> What's causing the SparseArrays error in CI?

#27469

**Comment by ViralBShah** [comment]
Bump.

**Comment by ararslan** [comment]
I've rebased against master and resolved the merge conflicts. Should be good to go once CI passes.

**Comment by pkofod** [comment]
Merge?

**Pull request merged**